pkg/controller/job/indexed_job_utils.go

```diff
@@ -246,6 +313,91 @@ func appendDuplicatedIndexPodsForRemoval(rm, left, pods []*v1.Pod, completions i
 	return appendPodsWithSameIndexForRemovalAndRemaining(rm, left, pods[firstRepeatPos:countLooped], lastIndex)
 }
 
+// getFailedIndexesMessage returns JobFailed condition message used when
+// the Job is marked failed due to presence of failed indexes.
+func getFailedIndexesMessage(failedIndexes orderedIntervals) string {
+	failedCount := failedIndexes.total()
+	indexes := getIndexesUpToCount(failedIndexes, maxIndexesInMessageString)
+	stringIndexes := make([]string, len(indexes))
+	for i, num := range indexes {
+		stringIndexes[i] = strconv.Itoa(num)
+	}
+	failedIndexesString := strings.Join(stringIndexes, ",")
+	if failedCount > maxIndexesInMessageString {
+		failedIndexesString = failedIndexesString + "..."
+	}
+	pluralSuffix := ""
+	if failedCount != 1 {
+		pluralSuffix = "es"
+	}
+	return fmt.Sprintf("Job has %d failed index%s (%s)", failedCount, pluralSuffix, failedIndexesString)
+}
+
+func getIndexesUpToCount(intervals orderedIntervals, maxCount int) []int {
+	var result []int
+	for _, interval := range intervals {
+		for index := interval.First; index <= interval.Last; index++ {
+			result = append(result, index)
+			if len(result) >= maxCount {
+				return result
+			}
+		}
+	}
+	return result
+}
+
+// getPodsWithDelayedDeletionPerIndex returns the pod which removal is delayed
+// in order to await for recreation. This map is used when BackoffLimitPerIndex
+// is enabled to delay pod finalizer removal, and thus pod deletion, until the
+// replacement pod is created. The pod deletion is delayed so that the
+// replacement pod can have the batch.kubernetes.io/job-index-failure-count
+// annotation set properly keeping track of the number of failed pods within
+// the index.
+func getPodsWithDelayedDeletionPerIndex(logger klog.Logger, job *batch.Job, pods, activePods []*v1.Pod, succeededIndexes, failedIndexes orderedIntervals, expectedRmFinalizers sets.Set[string]) map[int]*v1.Pod {
+	podsWithDelayedDeletionPerIndex := make(map[int]*v1.Pod)
+	// the failed pods corresponding to currently active indexes can be safely
+	// deleted as the failure count annotation is present in the currently
+	// active pods.
+	active := getIndexes(activePods)
+
+	failedPodsWithFinalizers := getValidPodsWithFilter(job, pods, nil, expectedRmFinalizers, func(p *v1.Pod) bool {
```

Comment:
Sort of, we have the `newFailedPods`, but they don't include pods which are already in the `uncounted`. Currently, I need to include such pods in the computation.

However, I could delay adding them into the `status.uncountedTerminatedPods` until they are ready to be replaced. This could also be beneficial because then the pods which are awaiting their recreation for 10s of backoff delay don't block the updates to  `status.uncountedTerminatedPods` - it would eliminate the new phenomenon that pods may spend a couple of seconds in `status.uncountedTerminatedPods`. WDYT?

Reply:
OTOH now thinking about it, exposing the failed pods in `status.uncountedTerminatedPods` has the benefit that the information that they failed is available in the API to other controllers, such as Kueue.

Reply:
So I think keeping as is makes sense from the API POV, but I have another idea how to optimize this function a little